### ISSUE TITLE
support CH functions for dates and times 

### ISSUE THREAD

**Issue by jinmingjian** [tensorbase/tensorbase#131]
- [] toTimeZone
- [] toYear
- [] toQuarter
- [] toMonth
- [] toDayOfYear
- [] toDayOfMonth
- [] toDayOfWeek
- [] toHour
- [] toMinute
- [] toSecond
- [] toUnixTimestamp
- [] toStartOfYear
- [] toStartOfISOYear
- [] toStartOfQuarter
- [] toStartOfMonth
- [] toMonday
- [] toStartOfWeek(t[,mode])
- [] toStartOfDay
- [] toStartOfHour
- [] toStartOfMinute
- [] toStartOfFiveMinute
- [] toStartOfTenMinutes
- [] toStartOfFifteenMinutes
- [] toStartOfInterval(time_or_data, INTERVAL x unit [, time_zone])
- [] toTime
- [] toRelativeYearNum
- [] toRelativeQuarterNum
- [] toRelativeMonthNum
- [] toRelativeWeekNum
- [] toRelativeDayNum
- [] toRelativeHourNum
- [] toRelativeMinuteNum
- [] toRelativeSecondNum
- [] toISOYear
- [] toISOWeek
- [] toWeek(date[,mode])
- [] toWeek(date,[, mode][,Timezone])
- [] toYearWeek(date[,mode])
- [] now
- [] today
- [] yesterday
- [] timeSlot
- [] toYYYYMM
- [] toYYYYMMDD
- [] toYYYYMMDDhhmmss
- [] addYears, addMonths, addWeeks, addDays, addHours, addMinutes, addSeconds, addQuarters
- [] subtractYears, subtractMonths, subtractWeeks, subtractDays, subtractHours, subtractMinutes, subtractSeconds, subtractQuarters
- [] dateDiff('unit', startdate, enddate,[timezone])
- [] timeSlots(StartTime, Duration,[, Size])
- [] formatDateTime(Time, Format[, Timezone])

**Comment by whjpji** [comment]
Here is a supporting plan based on the [Documentation of ClickHouse Dates and Times function](https://clickhouse.tech/docs/en/sql-reference/functions/date-time-functions/#timeslotsstarttime-duration-size).

1. First the most common date/time converting functions without timezone:
- [x] toYear
- [x] toQuarter
- [x] toMonth
- [x] toDayOfYear
- [x] toDayOfMonth
- [x] toDayOfWeek
- [x] toHour
- [x] toMinute
- [x] toSecond
- [ ] toTime
- [ ] now
- [ ] today
- [ ] yesterday
- [ ] toYYYYMM
- [ ] toYYYYMMDD
- [ ] toYYYYMMDDhhmmss

2. Then the date/time truncationss:
- [ ] toStartOfYear
- [ ] toStartOfISOYear
- [ ] toStartOfQuarter
- [ ] toStartOfMonth
- [ ] toMonday
- [ ] toStartOfWeek(t[,mode])
- [ ] toStartOfDay
- [ ] toStartOfHour
- [ ] toStartOfMinute
- [ ] toStartOfFiveMinute
- [ ] toStartOfTenMinutes
- [ ] toStartOfFifteenMinutes
- [ ] toStartOfInterval(time_or_data, INTERVAL x unit [, time_zone])

3. Next, the date/time additions/substractions:
- [ ] addYears
- [ ] addMonths
- [ ] addWeeks
- [ ] addDays
- [ ] addHours
- [ ] addMinutes
- [ ] addSeconds
- [ ] addQuarters
- [ ] subtractYears
- [ ] subtractMonths
- [ ] subtractWeeks
- [ ] subtractDays
- [ ] subtractHours
- [ ] subtractMinutes
- [ ] subtractSeconds
- [ ] subtractQuarters
- [ ] dateDiff('unit', startdate, enddate, [timezone])

4. Finally, other date/time functions including timezone:
- [ ] toTimeZone
- [ ] toUnixTimestamp
- [ ] toRelativeYearNum
- [ ] toRelativeQuarterNum
- [ ] toRelativeMonthNum
- [ ] toRelativeWeekNum
- [ ] toRelativeDayNum
- [ ] toRelativeHourNum
- [ ] toRelativeMinuteNum
- [ ] toRelativeSecondNum
- [ ] toISOYear
- [ ] toISOWeek
- [ ] toWeek(date[,mode])
- [ ] toWeek(date,[, mode][,Timezone])
- [ ] toYearWeek(date[,mode])
- [ ] timeSlots(StartTime, Duration,[, Size])
- [ ] formatDateTime(Time, Format[, Timezone])
- [ ] dateName
- [ ] FROM_UNIXTIME
- [ ] toModifiedJulianDay
- [ ] toModifiedJulianDayOrNull
- [ ] fromModifiedJulianDay
- [ ] fromModifiedJulianDayOrNull

**Comment by jinmingjian** [comment]
@whjpji thanks for detail sketch! 
Timezone is important. In fact, all date/time functions in CH are timezone-aware explicitly or implicitly. The timestamp itself is UTC based, but the interpretation to ts is the server's timezone based, at least for CH.  This means, a query to same physical data will return different results if the server' tz is different. 

If you are interesting to figure out a best tz hooking way for TB to DataFusion before we start more datetime functions? this problem is tracked by issue #27 in fact.

**Comment by whjpji** [comment]
@jinmingjian Yes I agree with you that timezone is important. Actually I often meet the timezone related issues when doing CRUD.

Let me see how TB and DF handles timezone, and I may adjust the priority of the support plans.

**Comment by jinmingjian** [comment]
@whjpji great, let me know if any problem.
